### ISSUE TITLE
Provide mechanism for managing resources across engines and executions

### ISSUE THREAD

**Issue by marcphilipp** [junit-team/junit5#2816]
The Platform should provide an extensible mechanism (`ServiceLoader`?) so that user code can create "closeable resources" that are injected as part of the context received by `TestEngine` implementations.

This feature is analogous to the `ExtensionContext.Store` concept from Jupiter but takes it to the Platform level. Moreover, it allows to reuse the same resource across multiple executions in a `LauncherSession`. We should advocate that resources stored in such a key-value store be immutable until their destruction. If such resources implement `AutoCloseable` they will get closed immediately before the `LauncherSession` they belong to is closed.

**Comment by PinkFloyded** [comment]
Hi @marcphilipp, if there is interest in this feature, I can take a stab at implementing this. Though I'm pretty new to the Junit codebase and would have to familiarize myself first.

**Comment by terencenmnpan** [comment]
> Hi @marcphilipp, if there is interest in this feature, I can take a stab at implementing this. Though I'm pretty new to the Junit codebase and would have to familiarize myself first.

I have interest, I'm interested in making an attempt at a change based on this for the cucumber jvm project.

**Comment by marcphilipp** [comment]
@PinkFloyded @terencenmnpan Thanks for offering! Has either of you already put some thought into where/how you'd envision the Launcher API to change? I think starting on this will require a core team decision first but a concrete proposal would speed things up.

**Comment by YongGoose** [comment]
@marcphilipp @terencenmnpan @PinkFloyded 

any updates?

**Comment by marcphilipp** [comment]
@YongGoose No, there are no updates. Are you interested in writing up a (short) proposal for discussion?

**Comment by YongGoose** [comment]
> [@YongGoose](https://github.com/YongGoose) No, there are no updates. Are you interested in writing up a (short) proposal for discussion?


Yes, I'm interested in writing up a proposal. However, it will take me a few days to conduct a thorough analysis and put together a detailed proposal.

**Comment by marcphilipp** [comment]
Sure, no worries! 👍

**Comment by YongGoose** [comment]
@marcphilipp 
I apologize for the delay in preparing the proposal.🥲
I’m still working through some thoughts and plan to start writing it after completing the following two tasks.
- https://github.com/junit-team/junit5/pull/4120
- https://github.com/junit-team/junit5/pull/4135

> I could work on the proposal at the same time, but I’ve decided to focus on the current tasks first to ensure the best quality for the project.

**Comment by marcphilipp** [comment]
One idea I had in the meantime would be to build upon #3064. We could create both a session-level and request-level instance of `NamespacedHierarchicalStore`. The former could be exposed to registered `LauncherSessionListeners` via `LauncherSession.getStore()`. The latter would be created in the `Launcher` for every `LauncherDiscoveryRequest` and use the session-level store as its parent. It would then be exposed to test engines via `EngineDiscoveryRequest` and to `TestExecutionListeners` via `TestPlan`. JUnit Jupiter (and other engines) would then use the request-level store as parent when creating the store for `JupiterEngineExtensionContext`.

**Comment by YongGoose** [comment]
@marcphilipp 

The comment was truly awesome 👍🏻 https://github.com/junit-team/junit5/issues/2816#issuecomment-2508195159

While writing the proposal, I initially considered a `single-instance` structure similar to [TestNG's IAttributes](https://github.com/testng-team/testng/blob/db6c5a9b0c827046f678d395b5178d55ac2a6d10/testng-core-api/src/main/java/org/testng/IAttributes.java#L11). 
However, I decided to retain the `hierarchical structure`, as it seems to be more effective for managing data.

So instead of completing the proposal, **I implemented the comment within my personal workspace.**

The work has been divided into several steps:

1. Designing a class to manage context at the platform level (https://github.com/YongGoose/junit5/commit/5bfe6ad6c7e0bc909842f6b1c293107382e674d6)
2. Adding a **session-level store** to `LauncherSession` (https://github.com/YongGoose/junit5/commit/7992634854d174fa98202872ffde1280f727c222)
3. Adding a **request-level store** to `Launcher`
4. Exposing the **request-level store** in `EngineDiscoveryRequest` and `TestPlan`
5. Utilizing the **request-level store** in `JupiterEngineExtensionContext`

---


If my implementation approach is suitable, I am willing to contribute. 

However, if the JUnit team has already decided to proceed with the implementation, 
I would like to participate (**even if just a little** 😁), in the discussions on the design direction and contribute to the project in any way I can.

**Comment by marcphilipp** [comment]
> 1\. Designing a class to manage context at the platform level ([YongGoose@5bfe6ad](https://github.com/YongGoose/junit5/commit/5bfe6ad6c7e0bc909842f6b1c293107382e674d6))

I think we only need a copy of Jupiter's `Namespace` class

> If my implementation approach is suitable, I am willing to contribute.
> 
> However, if the JUnit team has already decided to handle the implementation, I would like to participate (**even if just a little** 😁), in the discussions on the design direction and contribute to the project in any way I can.

Nobody from the team has already started on this so it's yours for the taking. However, may I kindly suggest we finish your other PRs first? 😅 

I think the steps you're proposing sound good! 👍 

We'll also have to think about how a Jupiter extension (or equivalent extension of another test engine) could store something in the session- or request-level store. We'll probably have to add `NamespacedHierarchicalStore.getParent() : Optional<NamespacedHierarchicalStore<N>>` for that and methods like `getSessionLevelStore() : Store` and `getRequestLevelStore(): Store` to `ExtensionContext`.

**Comment by YongGoose** [comment]
@marcphilipp 

While working on this issue, I created a Draft PR to ask questions along with the code if needed❕
- https://github.com/junit-team/junit5/pull/4252

I’m doing my best to understand and implement the issue, but there might be areas where I fall short.  
If there are any mistakes or areas for improvement, I’d greatly appreciate your feedback! 🙂
